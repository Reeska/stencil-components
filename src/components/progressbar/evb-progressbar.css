
:host {
  color: var(--evb-text-color);
  display: block;
  background-color: var(--evb-background-light);
  position: relative;
}

:host(.progress--with-text::after) {
  content: attr(data-progress) '%';
  display: block;
  font-size: 1em;
  left: 50%;
  position: absolute;
  top: 50%;
  transform: translate(-50%, -50%);
}

:host(.progress--gt-50) {
  color: var(--evb-text-color-invert);
}

.progress__inner {
  background-color: var(--evb-primary-color);
  height: 100%;
  transition: width .1s ease-out;
  width: 0;
}
