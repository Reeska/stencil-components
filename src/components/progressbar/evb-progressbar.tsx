import { Component, ComponentInterface, Event, EventEmitter, h, Prop, Watch } from '@stencil/core';

const HALFWAY_THERE = 50;
@Component({
  tag: 'evb-progressbar',
  styleUrl: 'evb-progressbar.css',
  shadow: true
})
export class EvbProgressBar implements ComponentInterface {

  /**
   * Progress percentage
   */
  @Prop() progress: number;
  /**
   * Show the progress as text in the progress bar
   */
  @Prop() text: boolean;
  /**
   * The height of the bar in pixels
   */
  @Prop() height = 24;

  @Event() completed: EventEmitter<void>;

  @Watch('progress')
  progressUpdate(current: number) {
    if (current >= 100) {
      this.completed.emit();
    }
  }

  hostData() {
    return {
      class: { 'progress--with-text': this.text, 'progress--gt-50': this.progress > HALFWAY_THERE },
      style: { height: `${this.height}px` },
      'data-progress': this.progress || 0
    };
  }
  /**
   * Renders the components
   */
  render() {
    const styles = {
      width: this.progress ? `${this.progress}%` : ''
    };

    return <div class='progress__inner' style={styles}></div>;
  }
}
