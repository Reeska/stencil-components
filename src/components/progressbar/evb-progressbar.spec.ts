import { EvbProgressBar } from './evb-progressbar';

describe('EvbProgressBar', () => {
  let component: EvbProgressBar;

  beforeEach(() => component = new EvbProgressBar());

  it('should render', () => {
    expect(component.render()).toMatchSnapshot();
  });

  describe('progress', () => {

    it('should have a default data-progress attribute on the host', () => {
      const hostData = component.hostData();
      expect(hostData['data-progress']).toBe(0);
    });

    it('should have data-progress attribute on the host which reflects the progress', () => {
      const expected = 88;
      component.progress = expected;
      const hostData = component.hostData();
      expect(hostData['data-progress']).toBe(expected);
    });

    it('should emit when progress is 100', () => {
      const emit = jest.fn();
      component.completed = { emit };

      // tslint:disable-next-line: no-magic-numbers
      component.progress = 99;

      expect(emit).not.toHaveBeenCalled();

      component.progress = 100;
      component.progressUpdate(component.progress);

      expect(emit).toHaveBeenCalled();
    });
  });
});
