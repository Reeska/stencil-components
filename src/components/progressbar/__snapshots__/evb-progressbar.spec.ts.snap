// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EvbProgressBar should render 1`] = `
Object {
  "$attrs$": Object {
    "class": "",
    "data-progress": 0,
    "style": Object {
      "height": "24px",
    },
  },
  "$children$": Array [
    Object {
      "$attrs$": Object {
        "class": "progress__inner",
        "style": Object {
          "width": "",
        },
      },
      "$children$": null,
      "$elm$": undefined,
      "$flags$": 0,
      "$key$": undefined,
      "$name$": undefined,
      "$tag$": "div",
    },
  ],
  "$elm$": undefined,
  "$flags$": 0,
  "$key$": undefined,
  "$name$": undefined,
  "$tag$": Object {},
}
`;
